Lumped parameters extraction and rewriting for polynomials

ggould-tri opened the pull request:
NOTE:  This PR will be reviewed offline next week.  It does not require review in github/reviewable for now.

This is the first of what will be several PRs to add system identification features to Drake.  This PR implements methods to identify lumped parameters in a polynomial (or set of polynomials) and to rewrite the polynomial in terms of those lumped parameters.

In addition it adds several necessary new methods to Polynomial, tests for those methods, and documentation a bug/shortcoming of Polynomial in the form of a disabled test.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2110)
<!-- Reviewable:end -->

RussTedrake commented:
took just a quick pass through this.  

---

Reviewed 7 of 8 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/util/CMakeLists.txt, line 18 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoGqbs2mzsagr5EJlS:-KFoGqbs2mzsagr5EJlT:1662367722) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/CMakeLists.txt#L18)):*
seems a bit odd to add SystemIdentification to drakePolynomial

---

*[drake/util/Polynomial.h, line 80 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoIwlNupsjy3eq5lpT:-KFoIwlNupsjy3eq5lpU:-954385619) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L80)):*
I would think adding DRAKEPOLYNOMIAL_EXPORT here would resolve your win32 issues below?

---

*[drake/util/Polynomial.h, line 203 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoHVexsVV5pLmf8aNb:-KFoHVexsVV5pLmf8aNc:-1134356805) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L203)):*
the method is called multivariateValue (but the doc says univariate).  

more fundamentally, i don't understand why you need to distinguish between the two.  

---

*[drake/util/SystemIdentification.h, line 33 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoKde1h4bJZeTs0lgy:-KFoKde1h4bJZeTs0lgz:1926035654) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L33)):*
i'm not a fan of the name "variables of interest".  it's too vague.

in my view it would be much more clean if you passed in the list of parameters here, instead of the non-parameters (which are often state, but i agree could be anything "of interest"... but the parameters can always be referred to clearly as "parameters").  

---

*[drake/util/test/testPolynomial.cpp, line 232 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoIUz-vtQMnGYA3qep:-KFoIUz-vtQMnGYA3qeq:2075623765) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testPolynomial.cpp#L232)):*
is the bug in the ascii output, or does it actually generate the wrong (or not sufficiently minimal) polynomial?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-:-KFoJNRUyT8zZOsM3kTo:-1577681912)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Review status: all files reviewed at latest revision, 47 unresolved discussions.

---

*[drake/util/Polynomial.cpp, line 111 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp-RaIAnVgyyCOrC4K:-KFp-RaIAnVgyyCOrC4L:2142477319) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.cpp#L111)):*
should be const ref.
Performance  issue:  loop could be map from var to power. 

---

*[drake/util/Polynomial.cpp, line 121 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp0D4eI1L0EVl4OqV4:-KFp0D4eI1L0EVl4OqV5:-498126719) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.cpp#L121)):*
Other does not express fact that its actually the divisor; consider rename to divisor. 

---

*[drake/util/Polynomial.cpp, line 125 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp-XYWdVOr9wtjxHtc:-KFp-XYWdVOr9wtjxHtd:814996193) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.cpp#L125)):*
should be const ref

---

*[drake/util/Polynomial.h, line 7 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFowueFe-nNFPoKbjdi:-KFowueFe-nNFPoKbjdj:-557834174) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L7)):*
these lines should be before Eigen includes, per coding standard about std includes.

---

*[drake/util/Polynomial.h, line 205 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoy6gofIsfxAZqU_Cc:-KFoy6gofIsfxAZqU_Cd:-1646836473) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L205)):*
x should be xs?

---

*[drake/util/Polynomial.h, line 212 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoyUTjfc62nLOjt7Qk:-KFoyUTjfc62nLOjt7Ql:-1104316164) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L212)):*
Name of multivariateValue is confusing; multivariateEvaluate or eval).

---

*[drake/util/Polynomial.h, line 223 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoyqWCyM7mRl718_Nc:-KFoyqWCyM7mRl718_Nd:-627237150) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L223)):*
Code style should not use C++ cast. Just say 0, do not cast (result of 3rd hour discussion). 

---

*[drake/util/Polynomial.h, line 228 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFozoc9_oP_neDaUlik:-KFozocA1gwdTFFRZ7SG:-1775122023) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L228)):*
Should not use C++ style of casts.  
Defect: using something that starts with MSVC.

---

*[drake/util/SystemIdentification.cpp, line 16 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp7NBLxh15DgDPSCTg:-KFp7NBLxh15DgDPSCTh:316047425) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L16)):*
PolyType::Monomial should just be Monomial (already defined).

---

*[drake/util/SystemIdentification.cpp, line 19 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp7Z2Ifn0r42fRzePJ:-KFp7Z2Ifn0r42fRzePK:-1839845002) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L19)):*
explain what this line is doing:  testing that term.var is a var of interest.

---

*[drake/util/SystemIdentification.cpp, line 23 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp7pyfndh3Zr-TezZF:-KFp7pyfndh3Zr-TezZG:-699943040) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L23)):*
This can return constants. Is that correct?  What if the for loop does nothing?

---

*[drake/util/SystemIdentification.cpp, line 30 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp8S50kZdK9Ug0C2OP:-KFp8S50kZdK9Ug0C2OQ:-466130920) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L30)):*
Not clear how this code implements the headers doc.  Is vars_of_interest  the right name. Not sure what it's trying to do. 

---

*[drake/util/SystemIdentification.cpp, line 84 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpAfe9sLiKW7fzH-Fd:-KFpAfe9sLiKW7fzH-Fe:62101582) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L84)):*
Hard to understand.  Please explain the reservation of 1000 names and why you do it in a comment.
3rd hours discussion:  Grant is trying to test if lump already exists.  Consider adding unique var type value function to Polynomial ?   

---

*[drake/util/SystemIdentification.cpp, line 85 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp9jRSfZPb_MOMmnli:-KFp9jRSfZPb_MOMmnlj:1829179744) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L85)):*
Magic number, 1000, and lump violates rule of 3.
Polynomiald could be PolyType?? 

---

*[drake/util/SystemIdentification.cpp, line 89 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpAFkxzz3OAj4WJ0wA:-KFpAFkxzz3OAj4WJ0wB:-916902615) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L89)):*
Clearer to say "lumped param output name".

---

*[drake/util/SystemIdentification.cpp, line 94 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpB8Q26fozoiXL3etc:-KFpB8Q35yYSsbomQOkn:-1179302016) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L94)):*
prefer MononialType.

---

*[drake/util/SystemIdentification.cpp, line 104 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpBnNrus2QMF-vOkQm:-KFpBnNrus2QMF-vOkQn:1189392604) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L104)):*
Performance problem?   Maybe 5 for loops. 

---

*[drake/util/SystemIdentification.cpp, line 121 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpCLjeVTBFPoGNRLh2:-KFpCLjf0cVl39sM5RWi:22481604) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L121)):*
So we need to say SystemIdentifixation<T>?

---

*[drake/util/SystemIdentification.cpp, line 140 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpCml3cj7Oll4Pn322:-KFpCml3cj7Oll4Pn323:-852098618) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L140)):*
auto lump_name_paid should be a const & ref

---

*[drake/util/SystemIdentification.cpp, line 151 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpD06sOa121QyUpYe5:-KFpD06sOa121QyUpYe6:1653503799) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L151)):*
PolyType::Monomial could be MonomialType

---

*[drake/util/SystemIdentification.cpp, line 155 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpDeLGoidz039MLZpI:-KFpDeLGoidz039MLZpJ:1845894655) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L155)):*
Comments to add:  what var is shadowing what.
3rd hour discussion. Refactor this to avoid complex data flow. Helper functions?  Helps document things also, to have a separate function. 

---

*[drake/util/SystemIdentification.cpp, line 171 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpE7MOf0cJO5Ilmh74:-KFpE7MPQjQa6fJ0rNn1:-972311187) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L171)):*
indentation defect

---

*[drake/util/SystemIdentification.cpp, line 176 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpEKn8I6JG7SsjmusV:-KFpEKn8I6JG7SsjmusW:-1429790706) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L176)):*
reason this could happen: doing rewrite with diff set of lumped params.   Please expand comment about when and why this happens.

---

*[drake/util/SystemIdentification.cpp, line 198 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpGJJy6QWHNxFcKcyG:-KFpGJJy6QWHNxFcKcyH:1302146560) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L198)):*
superfluous semicolons.

---

*[drake/util/SystemIdentification.cpp, line 199 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpG6SO5I8uVBbTwWi5:-KFpG6SP1jNTIDa2jJa5:275446379) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L199)):*
code style requires closing namespace comments. 

---

*[drake/util/SystemIdentification.h, line 1 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp4DvB8MktzWdTvd2V:-KFp4DvB8MktzWdTvd2W:600825593) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L1)):*
Filename should match Google style  of lowercase with underscores.
#define include guard should now be pragma once.

---

*[drake/util/SystemIdentification.h, line 22 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp53ajxz_AAXIt8CZM:-KFp53ajxz_AAXIt8CZN:1440983211) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L22)):*
This is using the term of art "system id" without references to explain that term. (e.g. web links).

---

*[drake/util/SystemIdentification.h, line 24 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp5X9THEApDH_YwLx5:-KFp5X9THEApDH_YwLx6:-771731444) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L24)):*
Why the underscore?  Why can't you template on coefficentType...

---

*[drake/util/SystemIdentification.h, line 42 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp5uOHOKlFjvha-KrO:-KFp5uOHOKlFjvha-KrP:-1394025278) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L42)):*
Please explain what is returned? Exact values in map. 
Consider only having plural version.

---

*[drake/util/SystemIdentification.h, line 89 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp6ORLQN_1kmk-NIkT:-KFp6ORLQN_1kmk-NIkU:-1555019198) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L89)):*
missing quote.
Please include an example.

---

*[drake/util/SystemIdentification.h, line 102 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp6X-7XMVaed7qBMsB:-KFp6X-7XMVaed7qBMsC:531948421) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L102)):*
please include example in comments.
Normalize?  Change to canonicalize. 

---

*[drake/util/test/CMakeLists.txt, line 42 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp3ukFZDqevX7r0SPF:-KFp3ukGpahmh-HYww8_:805659879) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/CMakeLists.txt#L42)):*
Consider combining blocks of GTEST_FOUND

---

*[drake/util/test/testPolynomial.cpp, line 1 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp0y2Hlt2WiPU55PCz:-KFp0y2I7puSxkhEJeds:-1567691774) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testPolynomial.cpp#L1)):*
not specific: include stream

---

*[drake/util/test/testPolynomial.cpp, line 196 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp1SiN_OIIEnybcr54:-KFp1SiN_OIIEnybcr55:-834705124) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testPolynomial.cpp#L196)):*
the test name does not need to start with "test", and that is inconsistent with other test names. True in 4 cases in this file.

---

*[drake/util/test/testPolynomial.cpp, line 278 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp3AkE30-gIQ-CLoo5:-KFp3AkE30-gIQ-CLoo6:164462890) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testPolynomial.cpp#L278)):*
map should be const

---

*[drake/util/test/testSystemIdentification.cpp, line 1 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpFU4tJZ3ZK4DAfcD-:-KFpFU4urSACZw9txp9r:-452419508) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L1)):*
No need to include iostream?

---

*[drake/util/test/testSystemIdentification.cpp, line 4 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpFfjf3_Fgz7-__kDR:-KFpFfjf3_Fgz7-__kDS:-1145408569) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L4)):*
Is testutil used?

---

*[drake/util/test/testSystemIdentification.cpp, line 12 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpGVOF4U-AChcICGiR:-KFpGVOF4U-AChcICGiS:1279916279) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L12)):*
remove "test" in front of testLumpedSingle

---

*[drake/util/test/testSystemIdentification.cpp, line 28 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpGql5RZqGdXtC_wiM:-KFpGql5RZqGdXtC_wiN:2037024254) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L28)):*
there is some better gtest syntax that will give you the contents... (says David)

---

*[drake/util/test/testSystemIdentification.cpp, line 86 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpHeBL6H4WoaY75Oea:-KFpHeBL6H4WoaY75Oeb:2090402954) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L86)):*
Need comment to explain loop.  Using  map to store values into map .. confusing.

---

*[drake/util/test/testSystemIdentification.cpp, line 96 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpIs3Vf2yFT7q5QnXU:-KFpIs3Vf2yFT7q5QnXV:-1714554840) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L96)):*
These tests do not adequately test the rewritten polynomial. Some commentators asked to check for actual return value. 

---

*[drake/util/test/testSystemIdentification.cpp, line 111 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpFwKE5iKcobG4kzW9:-KFpFwKE5iKcobG4kzWA:-394342765) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L111)):*
does not match actual namespace

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Review status: all files reviewed at latest revision, 47 unresolved discussions.

---

*[drake/util/Polynomial.cpp, line 111 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp-RaIAnVgyyCOrC4K:-KFpPLoh3E-F71IKsi7A:520878965) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.cpp#L111)):*
My comments today are the results of an in person Fagan code inspection with Sam, David German and Grant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 47 unresolved discussions.

---

*[drake/util/Polynomial.h, line 80 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoIwlNupsjy3eq5lpT:-KFtU9AjEajFsmE2iFo4:-535193530) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L80)):*
Tried that at https://github.com/RobotLocomotion/drake/pull/2110/commits/6047a1d3a3bdc06084071fd2a1780236179d308e and it did not work.  Absent a msvc linker expert, I have disabled the test on windows at https://github.com/RobotLocomotion/drake/pull/2110/commits/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4

---

*[drake/util/Polynomial.h, line 203 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoHVexsVV5pLmf8aNb:-KFtUc7k89zqW_rcR9eo:1497134975) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L203)):*
value() takes a single value, and so has the wrong type signature for evaluating multivariate polynomials.  Our coding style discourages overloading.  Hence the need for a new method (or for breaking all calls to the existing method, but that seemed extreme).

The docs are wrong and the name is unconventional; both will be corrected in an upcoming commit.

---

*[drake/util/SystemIdentification.h, line 33 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoKde1h4bJZeTs0lgy:-KFtVI40BUJGjjZYhn2u:2104250892) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L33)):*
I don't like 'state' as that assumes the equations of motion are in manipulator form, which they may not be; if they are not, then input is also going to be in the polynomial.

Specifying the parameters rather than the variables of interest is an interesting idea and I will try it and see how it looks.  I am a bit concerned that it is a lot easier to enumerate all of the state and input variables of a complex system beforehand than all of the parameters, since complex systems have a way of growing additional parameters as you debug them.

---

*[drake/util/test/testPolynomial.cpp, line 232 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoIUz-vtQMnGYA3qep:-KFtVrauJQ40UgQiZyoS:-581842429) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testPolynomial.cpp#L232)):*
The bug is in the actual Polynomial, which does not simplify the terms together; the test is written in terms of the ascii representation only because it is much easier to read that way.

I will file this as an issue when I land this PR (so the issue can refer to the disabled test).

I am unsure of whether the commutative case (the second block, below) should also be considered a bug because I don't know if you have an intent of supporting multiplication-noncommutative `CoefficientType`s in the future (matrices, quaternions, ...).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: 4 of 10 files reviewed at latest revision, 46 unresolved discussions.

---

*[drake/util/CMakeLists.txt, line 18 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoGqbs2mzsagr5EJlS:-KFtiLjrDOxUNEj_kTEW:2138525561) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/CMakeLists.txt#L18)):*
Yes, this is odd and will be changed in an upcoming PR (inter alia, it needs to move from util to solvers anyway)

---

*[drake/util/Polynomial.cpp, line 111 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp-RaIAnVgyyCOrC4K:-KFtijCHP2rpuW-NKLSb:69233899) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.cpp#L111)):*
Fixed in r3

---

*[drake/util/Polynomial.cpp, line 121 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp0D4eI1L0EVl4OqV4:-KFtinQH-AZj72iVG4-m:69233899) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.cpp#L121)):*
Fixed in r3

---

*[drake/util/Polynomial.cpp, line 125 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp-XYWdVOr9wtjxHtc:-KFtioUtyzmMZAWRdtg0:69233899) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.cpp#L125)):*
Fixed in r3

---

*[drake/util/Polynomial.h, line 7 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFowueFe-nNFPoKbjdi:-KFtiqy2F62U7CGcd5tO:69233899) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L7)):*
Fixed in r3

---

*[drake/util/Polynomial.h, line 203 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoHVexsVV5pLmf8aNb:-KFtiqH-xHOEO-J97OYP:69233899) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L203)):*
Fixed in r3

---

*[drake/util/Polynomial.h, line 205 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoy6gofIsfxAZqU_Cc:-KFtirtqHcJZutDSfPZl:1033084905) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L205)):*
Fixed in r3 (my Haskell background shows -- I naturally write "a collection of things like x" as "xs")

---

*[drake/util/Polynomial.h, line 212 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoyUTjfc62nLOjt7Qk:-KFtj-lPnzOX92f-f7yt:69233899) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L212)):*
Fixed in r3

---

*[drake/util/Polynomial.h, line 223 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoyqWCyM7mRl718_Nc:-KFtj0YeArx1K3QMFilN:69233899) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L223)):*
Fixed in r3

---

*[drake/util/Polynomial.h, line 228 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFozoc9_oP_neDaUlik:-KFtj1Nqe77hLnI5HIJb:69233899) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L228)):*
Fixed in r3

---

*[drake/util/test/testPolynomial.cpp, line 196 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp1SiN_OIIEnybcr54:-KFtj7HEhxuFS6Cr1gUQ:69233899) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testPolynomial.cpp#L196)):*
Fixed in r3

---

*[drake/util/test/testPolynomial.cpp, line 278 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp3AkE30-gIQ-CLoo5:-KFtj93Uj2FldAyB62j_:69233899) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testPolynomial.cpp#L278)):*
Fixed in r3

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: 2 of 10 files reviewed at latest revision, 35 unresolved discussions.

---

*[drake/util/SystemIdentification.cpp, line 16 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp7NBLxh15DgDPSCTg:-KFtoEzxhidOsqrecDnn:1736223991) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L16)):*
Fixed throughout in r4

---

*[drake/util/SystemIdentification.cpp, line 94 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpB8Q26fozoiXL3etc:-KFtoJn9Bv8eq_A5X-E9:1736223991) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L94)):*
Fixed throughout in r4

---

*[drake/util/SystemIdentification.cpp, line 104 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpBnNrus2QMF-vOkQm:-KFttRig9MrnZ7SwujvB:-1448814634) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L104)):*
commented in r4

---

*[drake/util/SystemIdentification.cpp, line 121 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpCLjeVTBFPoGNRLh2:-KFtoRRHXToomRwbqFOv:69233900) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L121)):*
Fixed in r4

---

*[drake/util/SystemIdentification.cpp, line 140 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpCml3cj7Oll4Pn322:-KFtolyUD9CWx749BPss:69233900) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L140)):*
Fixed in r4

---

*[drake/util/SystemIdentification.cpp, line 151 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpD06sOa121QyUpYe5:-KFtoeDKAbK4FuNGuKTt:1736223991) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L151)):*
Fixed throughout in r4

---

*[drake/util/SystemIdentification.cpp, line 171 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpE7MOf0cJO5Ilmh74:-KFtooZs5-h1Mz5lrYFd:69233900) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L171)):*
Fixed in r4

---

*[drake/util/SystemIdentification.cpp, line 198 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpGJJy6QWHNxFcKcyG:-KFtowEoeHk04Yoys5oS:69233900) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L198)):*
Fixed in r4

---

*[drake/util/SystemIdentification.cpp, line 199 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpG6SO5I8uVBbTwWi5:-KFtox9PliRs4zEk7k-4:69233900) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L199)):*
Fixed in r4

---

*[drake/util/SystemIdentification.h, line 1 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp4DvB8MktzWdTvd2V:-KFtpFOEgFav8SJB_D24:1947004449) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L1)):*
pragma once Fixed in r4
Filename will be fixed in upcoming PR that relocates this file.

---

*[drake/util/SystemIdentification.h, line 22 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp53ajxz_AAXIt8CZM:-KFtqcvxvaVol2hYK6Sg:2146250946) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L22)):*
Fixed in r4.

---

*[drake/util/SystemIdentification.h, line 24 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp5X9THEApDH_YwLx5:-KFtrW8alFuaJzo93coG:-1313814332) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L24)):*
No idea why this pattern was used in Polynomial; here it is simply cargo-cult.  Fixed in r4.

---

*[drake/util/SystemIdentification.h, line 89 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp6ORLQN_1kmk-NIkT:-KFtsDwQ6KuDqbfVpNgs:69233900) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L89)):*
Fixed in r4

---

*[drake/util/SystemIdentification.h, line 105 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFtpN4zCCHlpw5EqfcA:-KFtpN4zCCHlpw5EqfcB:-668754594) ([raw file](https://github.com/robotlocomotion/drake/blob/ad14f3d0f56a72ca07c3d3d177fc3be0ed28ccaa/drake/util/SystemIdentification.h#L105)):*
Superfluous semicolons; Fixed in r4

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: 2 of 10 files reviewed at latest revision, 21 unresolved discussions.

---

*[drake/util/SystemIdentification.cpp, line 19 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp7Z2Ifn0r42fRzePJ:-KFu63BLJOWG9caVancs:-1448814633) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L19)):*
commented in r5

---

*[drake/util/SystemIdentification.cpp, line 23 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp7pyfndh3Zr-TezZF:-KFu52fO9UEc5U9KzXds:-228365232) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L23)):*
To my great surprise, it is actually correct and necessary that this returns constant terms.  Documented this in the header at r5.

---

*[drake/util/SystemIdentification.cpp, line 30 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp8S50kZdK9Ug0C2OP:-KFu7R0q22YfPbxTwqkJ:-1448814633) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L30)):*
commented in r5

---

*[drake/util/SystemIdentification.cpp, line 84 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpAfe9sLiKW7fzH-Fd:-KFu0gj6aqA8NvcTCEO0:-646531457) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L84)):*
r5:  Cannot do this with unique value function in Polynomial because multiple polynomials defeat this.  Replaced this logic with "try until one is unique" (throws if none are unique before the large, undocumented maximum).

---

*[drake/util/SystemIdentification.cpp, line 85 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp9jRSfZPb_MOMmnli:-KFu1Iea_e9DqL0rB8Om:2146250977) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L85)):*
Fixed in r5.

---

*[drake/util/SystemIdentification.cpp, line 89 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpAFkxzz3OAj4WJ0wA:-KFu1UTBAJ0Ifzj2BGYl:1828245957) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L89)):*
Code is gone in r5

---

*[drake/util/SystemIdentification.h, line 42 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp5uOHOKlFjvha-KrO:-KFu8DXRpyQkDF39JRvs:558010320) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L42)):*
Comment expanded to include examples, r5.

Common-case calls in the tests got boilerplate-y when I removed the singular case, and the documentation wording was even less clear, so I'm going to take an author's-discretion WONTFIX on that one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

Lucy-tri commented:
Reviewed 7 of 8 files at r1, 1 of 1 files at r2, 1 of 6 files at r3, 1 of 2 files at r5.
Review status: 4 of 10 files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/util/Polynomial.h, line 223 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoyqWCyM7mRl718_Nc:-KFu95h3Vm4VwWXUXAM_:-976149510) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L223)):*
I'm still seeing: (ProductType)0;

---

*[drake/util/SystemIdentification.cpp, line 19 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp7Z2Ifn0r42fRzePJ:-KFuAjoizMIaVjsRTeOC:-556193219) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L19)):*
Thank you!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: 4 of 10 files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[drake/util/SystemIdentification.h, line 102 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp6X-7XMVaed7qBMsB:-KFu_WGFj-6EPY8gMAup:-1718541516) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L102)):*
documented and renamed, r6

---

*[drake/util/test/CMakeLists.txt, line 42 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp3ukFZDqevX7r0SPF:-KFuZRmzBZ8Ub_7rOxt1:657259160) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/CMakeLists.txt#L42)):*
Unified all of the various GTEST_FOUND blocks (r6)

---

*[drake/util/test/testPolynomial.cpp, line 1 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFp0y2Hlt2WiPU55PCz:-KFuDV-g2d5rwks90IIZ:1861525077) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testPolynomial.cpp#L1)):*
Fixed r6

---

*[drake/util/test/testSystemIdentification.cpp, line 1 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpFU4tJZ3ZK4DAfcD-:-KFuDZQrUltHsafGZVM7:1796880949) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L1)):*
fixed r6

---

*[drake/util/test/testSystemIdentification.cpp, line 4 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpFfjf3_Fgz7-__kDR:-KFuDcmzvvFQqTxc4mBB:1861525077) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L4)):*
Fixed r6

---

*[drake/util/test/testSystemIdentification.cpp, line 12 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpGVOF4U-AChcICGiR:-KFuDsiFs9Zqs8nNizah:1861525077) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L12)):*
Fixed r6

---

*[drake/util/test/testSystemIdentification.cpp, line 28 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpGql5RZqGdXtC_wiM:-KFuXDYmvUamemouBV-L:1869977082) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L28)):*
Unable to get this working; added TODO instead (r6)

---

*[drake/util/test/testSystemIdentification.cpp, line 86 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpHeBL6H4WoaY75Oea:-KFuYdunKAsj4jxU_qO-:531204998) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L86)):*
done in r6

---

*[drake/util/test/testSystemIdentification.cpp, line 96 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpIs3Vf2yFT7q5QnXU:-KFuYfOC9HR9sNQKEw4Y:261899539) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L96)):*
TODO comment in r6 (because additional test coverage from structural comparison is low compared to effort of writing the test).

---

*[drake/util/test/testSystemIdentification.cpp, line 111 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpFwKE5iKcobG4kzW9:-KFuDmvvoE-tDzcP8LMp:1861525077) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/test/testSystemIdentification.cpp#L111)):*
Fixed r6

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
FYI, I am pretty sure those linux matlab failures aren't mine, given I haven't touched any matlab code.

billhoffman commented:
Yes, they are not your fault.  If you look at the console output you can see that there are build failures.  Those issues  should be fixed by this PR: https://github.com/RobotLocomotion/drake/pull/2141.  Which was merged into master.  So, if you rebase this on master, it should clear up the matlab failures.

ggould-tri commented:
Review status: 2 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/util/SystemIdentification.cpp, line 155 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpDeLGoidz039MLZpI:-KFzCBRaoum2KAMXWqgZ:-1202136571) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L155)):*
Attempts to refactor this resulted large amounts of net code increase due to `std::tuple`/`std::get`/`std::tie` boilerplate.  So absent a net improvement from refactoring, I'm commenting it up.

---

*[drake/util/SystemIdentification.cpp, line 176 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFpEKn8I6JG7SsjmusV:-KFzEuOYr7raqPCtoOHd:1468741525) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.cpp#L176)):*
Commented.

---

*[drake/util/SystemIdentification.h, line 33 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoKde1h4bJZeTs0lgy:-KFzWRkqqM6TbKFLd_jp:618207024) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/SystemIdentification.h#L33)):*
I have rephrased the interfaces of the public methods in terms of `parameters` and the implementations and private methods (which are substantially simpler working with the set of non-parameter variables) I use the term 'active variables' (because of that old standard problem that we have no term for 'variables that vary, as opposed to variables that are constants).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 1 of 8 files at r1, 3 of 6 files at r3, 2 of 5 files at r6, 3 of 3 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/util/Polynomial.h, line 203 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KGCqIedHvmNf1TvV4Mh:-KGCqIee3EG0RhDgNfM2:1360640216) ([raw file](https://github.com/robotlocomotion/drake/blob/d8111d2ad5e13f89a6a9095ddcb73a9d619cb83e/drake/util/Polynomial.h#L203)):*
Should this still say univariate?

---

*[drake/util/Polynomial.h, line 226 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KGCqi66LkJoIc2VeAsi:-KGCqi67iigo7-wvqfPd:-634001559) ([raw file](https://github.com/robotlocomotion/drake/blob/d8111d2ad5e13f89a6a9095ddcb73a9d619cb83e/drake/util/Polynomial.h#L226)):*
Line length (Drake does not appear to have an exception to the 80 column limit in the C++ style guide..   though now I'm realizing that I'm just assuming Reviewable is 80 columns...)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/util/Polynomial.h, line 203 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KGCqIedHvmNf1TvV4Mh:-KGCswZMjZmQzCArMdIO:-2067064818) ([raw file](https://github.com/robotlocomotion/drake/blob/d8111d2ad5e13f89a6a9095ddcb73a9d619cb83e/drake/util/Polynomial.h#L203)):*
Fixed, r10

---

*[drake/util/Polynomial.h, line 226 \[r9\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KGCqi66LkJoIc2VeAsi:-KGCsszwfQYM5tcA8Qpf:1872702432) ([raw file](https://github.com/robotlocomotion/drake/blob/d8111d2ad5e13f89a6a9095ddcb73a9d619cb83e/drake/util/Polynomial.h#L226)):*
Fixed r10

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 9 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-:-KGCtXJ1nbQrWHgHj4Ew:1417088721)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Per chat, @david-german-tri was on deck to merge this.

david-german-tri commented:
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/util/Polynomial.h, line 223 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoyqWCyM7mRl718_Nc:-KGIH1S3hxG3jHHAErR-:1383152852) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L223)):*
does not appear resolved to me either

---

*[drake/util/SystemIdentification.cpp, line 129 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KGIFU1hdbUUZHipvS-l:-KGIFU1hdbUUZHipvS-m:1000167911) ([raw file](https://github.com/robotlocomotion/drake/blob/53827888d40ddc9c294331ad232feb0409cf5d4a/drake/util/SystemIdentification.cpp#L129)):*
Non-blocking: As discussed during the third hour, I still think lump variable allocation should be hidden inside a static method `PolyType::CreateLumpVariable` - i.e. `PolyType` itself should enforce that "lump" is a reserved name.   If you disagree feel free not to do it (or even file an issue) - I'm just posting this comment to make sure it didn't fall through the cracks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: 7 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/util/Polynomial.h, line 223 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KFoyqWCyM7mRl718_Nc:-KGIP5W8pgNTaoZFjCMi:-1343718700) ([raw file](https://github.com/robotlocomotion/drake/blob/ba65abe1ac621b1a7cb44e924f2d46f6d3841fc4/drake/util/Polynomial.h#L223)):*
Ah, there were actually three occurrences of this and I only fixed one.  Fixing.

---

*[drake/util/SystemIdentification.cpp, line 129 \[r10\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-KGIFU1hdbUUZHipvS-l:-KGISEq4ZZByuPdzMwOr:-1494657268) ([raw file](https://github.com/robotlocomotion/drake/blob/53827888d40ddc9c294331ad232feb0409cf5d4a/drake/util/SystemIdentification.cpp#L129)):*
I don't think Polynomial knowing about special variable names is a good idea, as the lumped variable logic will be moving out of util soon, and so should only use public methods of Polynomial.  Having a special magic entry point that allows you to use "magic" names is an unnecessary interface complication, and also creates an incompatibility between the C++ and matlab Polynomial APIs.  In addition, unique name creation needs to be enforced across an entire set of polynomials, not just one at a time, so uniqueness enforcement couldn't move into Polynomial anyway without serious API ugliness.

I did however factor the variable creation logic within SystemIdentification so that the uniqueness enforcement is in one, clearly marked place now rather than at the end of a long and tedious function where it might be overlooked.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 7 of 10 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-:-KGJPj09bLXX_VXN8gNS:1417088721)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 1 of 3 files at r11.
Review status: 8 of 10 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
Reviewed 1 of 8 files at r1, 3 of 6 files at r3, 2 of 5 files at r6, 1 of 3 files at r8, 2 of 3 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110)*
<!-- Sent from Reviewable.io -->

david-german-tri commented:
cpplint failure appears genuine, albeit trivial.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2110#-:-KGJQH2wSKq9VT4Hdi0K:-1503225676)*
<!-- Sent from Reviewable.io -->
